pandas/core/dtypes/concat.py

```diff
@@ -21,11 +26,78 @@
 from pandas.core.construction import array, ensure_wrapped_if_datetimelike
 
 
+class NullArrayProxy:
+    """
+    Proxy object for an all-NA array.
+
+    Only stores the length of the array, and not the dtype. The dtype
+    will only be known when actually concatenating (after determining the
+    common dtype, for which this proxy is ignored).
+    Using this object avoids that the internals/concat.py needs to determine
+    the proper dtype and array type.
+    """
+
+    ndim = 1
+
+    def __init__(self, n: int):
+        self.n = n
+
+    @property
+    def shape(self):
+        return (self.n,)
+
+
+def _array_from_proxy(arr, dtype: DtypeObj, fill_value=lib.no_default):
+    """
+    Helper function to create the actual all-NA array from the NullArrayProxy object.
+
+    Parameters
+    ----------
+    arr : NullArrayProxy
+    dtype : the dtype for the resulting array
+    fill_value : scalar NA-like value
+        By default uses the ExtensionDtype's na_value or np.nan. For numpy
+        arrays, this can be overridden to be something else (eg None).
+
+    Returns
+    -------
+    np.ndarray or ExtensionArray
+    """
+    if is_extension_array_dtype(dtype):
+        return dtype.construct_array_type()._from_sequence(
+            [dtype.na_value] * arr.n, dtype=dtype
+        )
+    elif is_datetime64_ns_dtype(dtype):
+        from pandas.core.arrays import DatetimeArray
+
+        return DatetimeArray._from_sequence([NaT] * arr.n, dtype=dtype)
+    elif is_timedelta64_ns_dtype(dtype):
+        from pandas.core.arrays import TimedeltaArray
+
+        return TimedeltaArray._from_sequence([NaT] * arr.n, dtype=dtype)
+    else:
+        if is_integer_dtype(dtype):
+            dtype = "float64"
+            fill_value = np.nan
+        elif is_bool_dtype(dtype):
+            dtype = object
+
+        if fill_value is lib.no_default:
+            fill_value = np.nan
```

Comment:
see internals.concat._dtype_to_na_value; should this ever be None instead of nan?

Reply:
> see internals.concat._dtype_to_na_value; should this ever be None instead of nan?

Here it sometimes needs to be None for datetimelike data. But that doesn't seem to be the use case in internals `_dtype_to_na_value`. Do you have any idea what kind of use case hits that?

Reply:
this code is outdated; is this question still relevant?

Reply:
Looking at that code in internals/concat.py, I don't think it is relevant here. The `has_none_blocks` logic basically ensures that `get_reindexed_values` just gets the block itneger values vs uses float as dtype and np.nan as fill value (the None itself is not used as actual fill_value, it's only used to pass as `upcasted_na` to `get_reindexed_values` to signal no upcasting is needed. 

But here, if we are in the NullArrayProxy, we already know that we have missing values to introduce.

Reply:
poor communication on my part.  "is this question still relevant?" was intended to be about your question "Do you have any idea what kind of use case hits that?", not my question in the comment before that

Reply:
> poor communication on my part. "is this question still relevant?" was intended to be about your question "Do you have any idea what kind of use case hits that?", not my question in the comment before that

OK ;) Then from my side this is resolved (AFAIU, the None in internals/concat serves a different purpose)